proc/futex.c

```diff
@@ -0,0 +1,178 @@
+{
+	u32 key, idx, i;
+	futex_sleepqueue_t *futex_sleepqueue = NULL;
+
+	key = address >> 3;
+	key ^= key >> FUTEX_SLEEPQUEUES_BITS;
+	idx = key & FUTEX_SLEEPQUEUES_MASK;
+
+	/* Find a taken slot with the same address using linear probing */
+	i = idx;
+	while (1) {
+		if (process->futex_sleepqueues[i].address == address) {
+			futex_sleepqueue = &process->futex_sleepqueues[i];
+			break;
+		}
+		i = (i + 1) % FUTEX_SLEEPQUEUES_SIZE;
+		if (i == idx) {
+			break;
+		}
+	}
+
+	return futex_sleepqueue;
+}
+
+static bool proc_futexUnwait(futex_sleepqueue_t *sq, futex_waitctx_t *wc)
+{
+	bool r;
+	spinlock_ctx_t sc;
+
+	hal_spinlockSet(&sq->spinlock, &sc);
+	r = atomic_load(&wc->thread) != NULL;
+	if (r) {
+		LIST_REMOVE(&sq->waitctxs, wc);
+	}
+	hal_spinlockClear(&sq->spinlock, &sc);
+	return r;
+}
+
+int proc_futexWait(_Atomic(u32) *address, u32 value, time_t timeout)
+{
+	spinlock_ctx_t ctx;
+	futex_sleepqueue_t *sleepqueue;
+	futex_waitctx_t wc;
+	thread_t *current;
+	int err = EOK;
+	time_t total_timeout = 0;
+
+	current = proc_current();
+
+	proc_lockSet(&current->process->lock);
+	sleepqueue = proc_getFutexSleepQueue(current->process, (addr_t)address);
+	if (sleepqueue == NULL) {
+		sleepqueue = proc_allocFutexSleepQueue(current->process, (addr_t)address);
+	}
+	if (sleepqueue == NULL) {
+		proc_lockClear(&current->process->lock);
+		return -ENOMEM;
+	}
+	proc_lockClear(&current->process->lock);
+
+	atomic_store(&wc.thread, current);
+	hal_spinlockSet(&sleepqueue->spinlock, &ctx);
+	LIST_ADD(&sleepqueue->waitctxs, &wc);
+	hal_spinlockClear(&sleepqueue->spinlock, &ctx);
+
+	if (atomic_load(address) != value) {
+		err = -EAGAIN;
+		goto done;
+	}
+
+
+	if (atomic_load(&wc.thread) != NULL) {
+		if (timeout > 0) {
+			proc_gettime(&total_timeout, NULL);
+			total_timeout += timeout;
+		}
+
+		hal_spinlockSet(&sleepqueue->spinlock, &ctx);
+		err = proc_threadWaitInterruptible(&sleepqueue->threads, &sleepqueue->spinlock, total_timeout, &ctx);
+		hal_spinlockClear(&sleepqueue->spinlock, &ctx);
+	}
+
+	if (err != EOK || atomic_load(&wc.thread) != NULL) {
+		if (proc_futexUnwait(sleepqueue, &wc) == 0) {
+			err = EOK;
+		}
+	}
+	return err;
+done:
+	if (atomic_load(&wc.thread) != NULL) {
+		proc_futexUnwait(sleepqueue, &wc);
+	}
+	return err;
+}
+
+int proc_futexWakeup(_Atomic(u32) *address, u32 n_threads)
+{
+	futex_sleepqueue_t *sleepqueue;
+	futex_waitctx_t *wc;
+	thread_t *current;
+	int i = 0, woken = 0;
+	spinlock_ctx_t ctx;
+	thread_t *tmp;
+
+	if (n_threads == 0) {
+		return 0;
+	}
+
+	current = proc_current();
+
+	proc_lockSet(&current->process->lock);
+	sleepqueue = proc_getFutexSleepQueue(current->process, (addr_t)address);
+	if (sleepqueue == NULL) {
+		proc_lockClear(&current->process->lock);
+		return 0;
+	}
+	proc_lockClear(&current->process->lock);
+
+	hal_spinlockSet(&sleepqueue->spinlock, &ctx);
+
+	while ((wc = sleepqueue->waitctxs) != NULL) {
+		LIST_REMOVE(&sleepqueue->waitctxs, wc);
```

Comment:
You're assuming `wc->thread` is the same as the thread being woken up. Why is that?